### PR TITLE
H-2495: Create layer to translate LLM calls to Anthropic/OpenAI API calls

### PR DESCRIPTION
**Pull request by benwerner01**
## 🌟 What is the purpose of this PR?

This PR replaces the existing `getOpenAiResponse` method with a `getLlmResponse` method, which allows for switching between **Anthropic** and **OpenAI** models.

To facilitate easier switching between LLM providers, the `getLlmResponse` method:
- unifies the `tools` input parameter
- adds support for a `sanitizeInputBeforeValidation` method in tool definitions, to handle common mistakes the LLM may make when producing inputs
- parses the tool calls in a unified `parsedToolCalls` output
- unifies the values of the Anthropic `stop_reason` and OpenAi `finish_reason` in a single `stopReason` output

In addition, internally it handles the following (with retry logic):
- missing function calls, when the stop reason is `tool_call`
- in the case of OpenAI tool calls, un-parsable input arguments for a tool
- invalid tool call arguments (arguments that don't match the JSON schema)

## 🔗 Related links

- H-2495

## 🔍 What does this change?

- replaces the `getOpenAiResponse` with the `getLlmResponse` method

### 🚢 Has this modified a publishable library?

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] do not affect the execution graph

## 🐾 Next steps

- add unified input/output format for `messages`, so that these don't have to be revised by the caller when switching between Anthropic and OpenAi models (H-2527)

## 🛡 What tests cover this?

Manual testing.

## ❓ How to test this?

Test whether the AI related functionality works as it did previously, for example by visiting the `/ai` page and starting a research task.

## 📹 Demo

image

### CODE REVIEW

File: apps/hash-ai-worker-ts/src/activities/shared/get-llm-response.ts [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+    retryMessages: OpenAiLlmParams["messages"];
+  }): Promise<LlmResponse<OpenAiLlmParams>> => {
+    if (retryCount > maxRetryCount) {
+      return {
+        status: "exceeded-maximum-retries",
+      };
+    }
+
+    logger.debug(
+      `Retrying OpenAi call with the following retry messages: ${stringify(retryParams.retryMessages)}`,
```
Comment (nonparibus) [inline apps/hash-ai-worker-ts/src/activities/shared/get-llm-response.ts:302]:
```suggestion
      `Retrying OpenAI call with the following retry messages: ${stringify(retryParams.retryMessages)}`,
```

File: apps/hash-ai-worker-ts/src/activities/shared/get-llm-response.ts [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+    case "stop":
+      return "stop";
+    case "tool_calls":
+      return "tool_use";
+    case "length":
+      return "length";
+    case "content_filter":
+      return "content_filter";
+    default:
+      throw new Error(`Unexpected OpenAi finish reason: ${finishReason}`);
```
Comment (nonparibus) [inline apps/hash-ai-worker-ts/src/activities/shared/get-llm-response.ts:76]:
```suggestion
      throw new Error(`Unexpected OpenAI finish reason: ${finishReason}`);
```

Comment (benwerner01) [reply apps/hash-ai-worker-ts/src/activities/shared/get-llm-response.ts]: Thanks, addressed in https://github.com/hashintel/hash/pull/4318/commits/f97fda48b5855f1725daa645196a3da68373818b

File: apps/hash-ai-worker-ts/src/activities/shared/get-llm-response.ts (hunk 1 above)
Comment (benwerner01) [reply apps/hash-ai-worker-ts/src/activities/shared/get-llm-response.ts]: Thanks, addressed in https://github.com/hashintel/hash/pull/4318/commits/f97fda48b5855f1725daa645196a3da68373818b

File: apps/hash-external-services/docker-compose.ai.yml [hunk 3]
```diff
@@ -19,6 +19,7 @@ services:
       HASH_TEMPORAL_SERVER_HOST: "http://temporal"
       HASH_TEMPORAL_SERVER_PORT: "${HASH_TEMPORAL_SERVER_PORT}"
       OPENAI_API_KEY: "${OPENAI_API_KEY}"
+      ANTHROPIC_API_KEY: "${ANTHROPIC_API_KEY}"
```
Comment (CiaranMn) [inline apps/hash-external-services/docker-compose.ai.yml:22]: Can you add this to the `README.md` for `hash-ai-worker-ts` please?

**Review by CiaranMn — COMMENTED** [review]
Code looks good, just one comment on docs – I can approve once Playwright is passing

File: apps/hash-external-services/docker-compose.ai.yml (hunk 3 above)
Comment (benwerner01) [reply apps/hash-external-services/docker-compose.ai.yml:22]: Yes, have added in https://github.com/hashintel/hash/pull/4318/commits/78c33786919f4145861d0a2b9dc03ee98f3e7ee8

**Review by CiaranMn — APPROVED** [review]

**Review by nonparibus — DISMISSED** [review]

**Review by CiaranMn — APPROVED** [review]

**Pull request merged**